File: backend/app/routers/goals.py
"""目标管理 API 路由"""

from fastapi import APIRouter, Depends, HTTPException, Query, Response

from app.api.schemas.goal import (
    GoalCreate,
    GoalUpdate,
    GoalResponse,
    GoalListResponse,
    GoalDetailResponse,
    GoalEntryCreate,
    GoalEntryLinkResponse,
    GoalEntryListResponse,
    ChecklistItemToggle,
    ProgressSummaryResponse,
)
from app.routers.deps import get_current_user, get_goal_service
from app.models.user import User

router = APIRouter(prefix="/goals", tags=["goals"])


@router.post("", response_model=GoalResponse, status_code=201)
async def create_goal(
    request: GoalCreate,
    user: User = Depends(get_current_user),
):
    """创建目标"""
    service = get_goal_service()
    result, status_code, message = await service.create_goal(request, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return result


@router.get("", response_model=GoalListResponse)
async def list_goals(
    status: str = Query(None, description="按状态过滤: active/completed/abandoned"),
    limit: int = Query(20, ge=1, le=100, description="每页数量"),
    user: User = Depends(get_current_user),
):
    """列出目标"""
    service = get_goal_service()
    results, status_code, message = await service.list_goals(user.id, status=status, limit=limit)
    return GoalListResponse(goals=results)


# NOTE: progress-summary 必须在 {goal_id} 路由之前定义，避免路径冲突
@router.get("/progress-summary", response_model=ProgressSummaryResponse)
async def get_progress_summary(
    period: str = Query(None, description="时间范围，如 week/month"),
    user: User = Depends(get_current_user),
):
    """获取进度汇总"""
    service = get_goal_service()
    result, status_code, message = await service.get_progress_summary(user.id, period=period)
    return result


@router.get("/{goal_id}", response_model=GoalDetailResponse)
async def get_goal(
    goal_id: str,
    user: User = Depends(get_current_user),
):
    """获取目标详情"""
    service = get_goal_service()
    result, status_code, message = await service.get_goal(goal_id, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return result


@router.put("/{goal_id}", response_model=GoalResponse)
async def update_goal(
    goal_id: str,
    request: GoalUpdate,
    user: User = Depends(get_current_user),
):
    """更新目标"""
    service = get_goal_service()
    result, status_code, message = await service.update_goal(goal_id, request, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return result


@router.delete("/{goal_id}")
async def delete_goal(
    goal_id: str,
    user: User = Depends(get_current_user),
):
    """删除目标（仅 abandoned 状态可删除）"""
    service = get_goal_service()
    result, status_code, message = await service.delete_goal(goal_id, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return {"message": message}


# === 条目关联 ===

@router.post("/{goal_id}/entries", response_model=GoalEntryLinkResponse, status_code=201)
async def link_entry(
    goal_id: str,
    request: GoalEntryCreate,
    user: User = Depends(get_current_user),
):
    """关联条目到目标"""
    service = get_goal_service()
    result, status_code, message = await service.link_entry(goal_id, request.entry_id, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return result


@router.delete("/{goal_id}/entries/{entry_id}", status_code=204)
async def unlink_entry(
    goal_id: str,
    entry_id: str,
    user: User = Depends(get_current_user),
):
    """取消关联条目"""
    service = get_goal_service()
    result, status_code, message = await service.unlink_entry(goal_id, entry_id, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return Response(status_code=204)


@router.get("/{goal_id}/entries", response_model=GoalEntryListResponse)
async def list_goal_entries(
    goal_id: str,
    user: User = Depends(get_current_user),
):
    """列出目标关联的条目"""
    service = get_goal_service()
    result, status_code, message = await service.list_goal_entries(goal_id, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return GoalEntryListResponse(entries=result)


# === Checklist ===

@router.patch("/{goal_id}/checklist/{item_id}", response_model=GoalResponse)
async def toggle_checklist_item(
    goal_id: str,
    item_id: str,
    user: User = Depends(get_current_user),
):
    """切换检查清单项的勾选状态"""
    service = get_goal_service()
    result, status_code, message = await service.toggle_checklist_item(goal_id, item_id, user_id=user.id)
    if result is None:
        raise HTTPException(status_code=status_code, detail=message)
    return result

File: e2e/tests/helpers/goals.ts
import { APIRequestContext } from '@playwright/test';

/**
 * Goals API 辅助函数
 *
 * 封装目标管理 CRUD 操作，所有调用走前端 vite proxy（/api/* → 后端）。
 * 需要传入已认证的 APIRequestContext（带 Authorization header）。
 */

/** 创建带认证头的请求选项 */
function authHeaders(token?: string): Record<string, string> {
  if (!token) return {};
  return { Authorization: `Bearer ${token}` };
}

export type MetricType = 'count' | 'checklist' | 'tag_auto';

export interface CreateGoalParams {
  title: string;
  metric_type: MetricType;
  target_value: number;
  auto_tags?: string[];
  checklist_items?: string[];
  start_date?: string;
  end_date?: string;
  description?: string;
}

/**
 * 创建目标
 * POST /api/goals
 */
export async function createGoal(
  request: APIRequestContext,
  params: CreateGoalParams,
  token?: string
) {
  const resp = await request.post('/api/goals', {
    headers: authHeaders(token),
    data: params,
  });
  return resp;
}

/**
 * 列出目标
 * GET /api/goals?status=active&limit=20
 */
export async function listGoals(
  request: APIRequestContext,
  token?: string,
  options?: { status?: 'active' | 'completed' | 'abandoned'; limit?: number }
) {
  const params = new URLSearchParams();
  if (options?.status) params.set('status', options.status);
  if (options?.limit) params.set('limit', String(options.limit));
  const qs = params.toString();
  const url = qs ? `/api/goals?${qs}` : '/api/goals';

  const resp = await request.get(url, {
    headers: authHeaders(token),
  });
  return resp;
}

/**
 * 获取目标详情（含 linked_entries_count）
 * GET /api/goals/{id}
 */
export async function getGoal(
  request: APIRequestContext,
  goalId: string,
  token?: string
) {
  const resp = await request.get(`/api/goals/${encodeURIComponent(goalId)}`, {
    headers: authHeaders(token),
  });
  return resp;
}

/**
 * 更新目标
 * PUT /api/goals/{id}
 */
export async function updateGoal(
  request: APIRequestContext,
  goalId: string,
  data: Record<string, unknown>,
  token?: string
) {
  const resp = await request.put(`/api/goals/${encodeURIComponent(goalId)}`, {
    headers: authHeaders(token),
    data,
  });
  return resp;
}

/**
 * 删除目标（仅 abandoned 可删）
 * DELETE /api/goals/{id}
 */
export async function deleteGoal(
  request: APIRequestContext,
  goalId: string,
  token?: string
) {
  const resp = await request.delete(`/api/goals/${encodeURIComponent(goalId)}`, {
    headers: authHeaders(token),
  });
  return resp;
}

/**
 * 关联条目到目标
 * POST /api/goals/{id}/entries
 */
export async function linkEntry(
  request: APIRequestContext,
  goalId: string,
  entryId: string,
  token?: string
) {
  const resp = await request.post(`/api/goals/${encodeURIComponent(goalId)}/entries`, {
    headers: authHeaders(token),
    data: { entry_id: entryId },
  });
  return resp;
}

/**
 * 取消关联条目
 * DELETE /api/goals/{id}/entries/{entry_id}
 */
export async function unlinkEntry(
  request: APIRequestContext,
  goalId: string,
  entryId: string,
  token?: string
) {
  const resp = await request.delete(`/api/goals/${encodeURIComponent(goalId)}/entries/${encodeURIComponent(entryId)}`, {
    headers: authHeaders(token),
  });
  return resp;
}

/**
 * 获取目标关联的条目列表
 * GET /api/goals/{id}/entries
 */
export async function listGoalEntries(
  request: APIRequestContext,
  goalId: string,
  token?: string
) {
  const resp = await request.get(`/api/goals/${encodeURIComponent(goalId)}/entries`, {
    headers: authHeaders(token),
  });
  return resp;
}

/**
 * 切换检查清单项
 * PATCH /api/goals/{id}/checklist/{item_id}
 */
export async function toggleChecklist(
  request: APIRequestContext,
  goalId: string,
  itemId: string,
  token?: string
) {
  const resp = await request.patch(`/api/goals/${encodeURIComponent(goalId)}/checklist/${encodeURIComponent(itemId)}`, {
    headers: authHeaders(token),
  });
  return resp;
}

/**
 * 获取进度概览
 * GET /api/goals/progress-summary?period=weekly
 */
export async function getProgressSummary(
  request: APIRequestContext,
  token?: string,
  period?: string
) {
  const params = new URLSearchParams();
  if (period) params.set('period', period);
  const qs = params.toString();
  const url = qs ? `/api/goals/progress-summary?${qs}` : '/api/goals/progress-summary';

  const resp = await request.get(url, {
    headers: authHeaders(token),
  });
  return resp;
}
